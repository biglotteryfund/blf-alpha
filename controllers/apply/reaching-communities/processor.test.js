/* eslint-env jest */
'use strict';
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const nodemailer = require('nodemailer');
const { get, map } = require('lodash');
const jsdom = require('jsdom');
const { JSDOM } = jsdom;

const { flattenFormData, stepsWithValues } = require('../helpers');
const form = require('./form-model');
const processor = require('./processor');

function mockProcessor() {
    const langData = fs.readFileSync(path.resolve(__dirname, '../../../config/locales/en.yml'), 'utf-8');
    const enCopy = yaml.safeLoad(langData);
    const formCopy = get(enCopy, form.lang);

    const mockFormData = {
        'step-1': {
            'your-idea': 'Test idea'
        },
        'step-2': {
            'location': 'South West',
            'project-location': 'London'
        },
        'step-3': {
            'organisation-name': 'Test Organisation',
            'additional-organisations': ''
        },
        'step-4': {
            'first-name': 'Example',
            'last-name': 'Person',
            'email': 'example@example.com',
            'phone-number': '03454102030'
        }
    };

    const mockTransport = nodemailer.createTransport({
        jsonTransport: true
    });

    const processorOptions = {
        data: flattenFormData(mockFormData),
        copy: formCopy,
        stepsWithValues: stepsWithValues(form.steps, mockFormData)
    };

    return processor(processorOptions, mockTransport);
}

describe('reaching communities processor', () => {
    test('customer email', async () => {
        const result = await mockProcessor();

        const customerEmail = result[0];
        const customerEmailMessage = JSON.parse(customerEmail.message);
        expect(customerEmailMessage.text).toMatchSnapshot();

        const dom = new JSDOM(customerEmailMessage.html);

        const summaryEl = dom.window.document.querySelector('.summary');

        expect(map(summaryEl.querySelectorAll('h4'), h => h.textContent)).toEqual([
            'Your idea',
            'Project location',
            'Your organisation',
            'Your details'
        ]);

        expect(map(summaryEl.querySelectorAll('h5'), h => h.textContent)).toEqual([
            'Briefly explain your idea and why it’ll make a difference',
            'Select all regions that apply',
            'Project location',
            'Legal name',
            'First name',
            'Last name',
            'Email address',
            'Phone number'
        ]);

        expect(map(summaryEl.querySelectorAll('h5 + p'), h => h.textContent)).toEqual([
            'Test idea',
            'South West',
            'London',
            'Test Organisation',
            'Example',
            'Person',
            'example@example.com',
            '03454102030'
        ]);
    });

    test('internal email', async () => {
        const result = await mockProcessor();

        const internalEmail = result[1];
        const internalEmailMessage = JSON.parse(internalEmail.message);
        expect(internalEmailMessage.text).toMatchSnapshot();

        const dom = new JSDOM(internalEmailMessage.html);

        const summaryEl = dom.window.document.querySelector('.summary');

        expect(map(summaryEl.querySelectorAll('h4'), h => h.textContent)).toEqual([
            'Your details',
            'Your organisation',
            'Project location',
            'Your idea'
        ]);

        expect(map(summaryEl.querySelectorAll('h5'), h => h.textContent)).toEqual([
            'First name',
            'Last name',
            'Email address',
            'Phone number',
            'Legal name',
            'Select all regions that apply',
            'Project location',
            'Briefly explain your idea and why it’ll make a difference'
        ]);

        expect(map(summaryEl.querySelectorAll('h5 + p'), h => h.textContent)).toEqual([
            'Example',
            'Person',
            'example@example.com',
            '03454102030',
            'Test Organisation',
            'South West',
            'London',
            'Test idea'
        ]);
    });
});
