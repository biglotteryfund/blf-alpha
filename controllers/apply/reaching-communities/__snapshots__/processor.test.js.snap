// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reaching communities processor customer email 1`] = `
" Thank you for getting in touch with us.

You will see a copy of the information you submitted below. This will be forwarded to one of our funding officers who will get in
touch within fifteen working days to let you know if we can help.

In the meantime, you can find out more about our funding and how we can support your community on our website: 
www.tnlcommunityfund.org.uk/england [https://www.tnlcommunityfund.org.uk/england]

Thanks,

The National Lottery Community Fund

YOUR INFORMATION
YOUR IDEA
BRIEFLY EXPLAIN YOUR IDEA AND WHY IT’LL MAKE A DIFFERENCE
Test idea

PROJECT LOCATION
SELECT ALL REGIONS THAT APPLY
South West

PROJECT LOCATION
London

YOUR ORGANISATION
LEGAL NAME
Test Organisation

YOUR DETAILS
FIRST NAME
Example

LAST NAME
Person

EMAIL ADDRESS
example@example.com

PHONE NUMBER
03454102030

DATA PROTECTION
The National Lottery Community Fund is a public body with a duty to distribute National Lottery and other money in grants for good
causes. We use the personal data you provide, such as contact details for individuals at your organisation, to help you apply for
a grant and to assess your application. We may share your personal data with organisations which help us with our grant making
activities or others which have a legitimate interest in our work or have funded your grant. We will only share personal data
which they need to carry out their work and subject to appropriate safety measures. Our Data Protection and Privacy Notice gives
more information about how we store and use personal data and the lawful basis for this. Please read the full Notice which is
published on our website at www.tnlcommunityfund.org.uk/data-protection [https://www.tnlcommunityfund.org.uk/data-protection] or
contact us to request a hard copy. The Notice may be updated from time to time."
`;

exports[`reaching communities processor internal email 1`] = `
" A CUSTOMER HAS SUBMITTED AN IDEA FROM THE REACHING COMMUNITIES & PARTNERSHIPS WEB FORM.
The details they provided are included below:

CUSTOMER INFORMATION
YOUR DETAILS
FIRST NAME
Example

LAST NAME
Person

EMAIL ADDRESS
example@example.com

PHONE NUMBER
03454102030

YOUR ORGANISATION
LEGAL NAME
Test Organisation

PROJECT LOCATION
SELECT ALL REGIONS THAT APPLY
South West

PROJECT LOCATION
London

YOUR IDEA
BRIEFLY EXPLAIN YOUR IDEA AND WHY IT’LL MAKE A DIFFERENCE
Test idea"
`;
